velox/exec/Window.cpp

```diff
@@ -0,0 +1,465 @@
+    allFuncsRawPeerStartBuffer[w] = rawPeerStartBuffer;
+    allFuncsRawPeerEndBuffer[w] = rawPeerEndBuffer;
+    allFuncsRawFrameStartBuffer[w] = rawFrameStartBuffer;
+    allFuncsRawFrameEndBuffer[w] = rawFrameEndBuffer;
+  }
+
+  size_t firstPartitionRow = partitionStartRows_[currentPartitionIndex_];
+  size_t lastPartitionRow = partitionStartRows_[currentPartitionIndex_ + 1] - 1;
+  for (int i = startRow, j = 0; i < endRow; i++, j++) {
+    // Compute the next peerStart and peerEnd rows (if this is the first row
+    // of the partition or if we are past previous peerGroup).
+    if (i == firstPartitionRow || i >= peerEndRow_) {
+      peerStartRow_ = i;
+      peerEndRow_ = i;
+      while (peerEndRow_ <= lastPartitionRow) {
+        if (peerCompare(
+                returningRows_[peerStartRow_], returningRows_[peerEndRow_])) {
+          break;
+        }
+        peerEndRow_++;
+      }
+    }
+    for (int w = 0; w < numFuncs; w++) {
+      // The peer and frame values set in the WindowFunction::apply buffers
+      // are the offsets within the current partition, so offset all row
+      // indexes correctly.
+      allFuncsRawPeerStartBuffer[w][j] = peerStartRow_ - firstPartitionRow;
+      allFuncsRawPeerEndBuffer[w][j] = peerEndRow_ - 1 - firstPartitionRow;
+
+      auto frameEndPoints = findFrameEndPoints(
+          w, partitionStartRows_[currentPartitionIndex_], endRow, i);
+      VELOX_CHECK_EQ(frameEndPoints.second, i);
+      allFuncsRawFrameStartBuffer[w][j] =
+          frameEndPoints.first - firstPartitionRow;
+      allFuncsRawFrameEndBuffer[w][j] =
+          frameEndPoints.second - firstPartitionRow;
+    }
+  }
+  // Invoke the apply method for the WindowFunctions
+  for (int w = 0; w < numFuncs; w++) {
+    windowFunctions_[w]->apply(
+        allFuncsPeerStartBuffer[w],
+        allFuncsPeerEndBuffer[w],
+        allFuncsFrameStartBuffer[w],
+        allFuncsFrameEndBuffer[w],
+        startRow - firstPartitionRow,
+        bufferIndex,
+        windowFunctionOutputs[w]);
+  }
+
+  numRowsApplied_ += numRows;
+  if (endRow == partitionStartRows_[currentPartitionIndex_ + 1]) {
+    currentPartitionIndex_++;
+  }
+}
+
+RowVectorPtr Window::getOutput() {
+  if (finished_ || !noMoreInput_ || returningRows_.size() == numRowsApplied_) {
+    return nullptr;
+  }
+
+  size_t rowsForOutput =
+      std::min(numRowsPerOutput_, returningRows_.size() - numRowsApplied_);
+  auto result = std::dynamic_pointer_cast<RowVector>(
+      BaseVector::create(outputType_, rowsForOutput, operatorCtx_->pool()));
+
+  // Set values of all output columns corresponding to the input columns.
+  // ExtractColumn might be doing a copy. Try to avoid that.
+  for (int i = 0; i < inputColumnsSize_; ++i) {
+    data_->extractColumn(
+        returningRows_.data() + numRowsApplied_,
+        rowsForOutput,
+        i,
+        result->childAt(i));
+  }
+
+  std::vector<VectorPtr> windowFunctionOutputs;
```

Comment:
No need for this vector. Use result instead.

Reply:
I have used this vector in intermediate computations.